### PR TITLE
PP-719 Logging when authentication fails

### PR DESCRIPTION
**Pull request by imranbohoran**
## WHAT

_A brief description of the pull request:_
- We don't have useful logging to indicate if smartpay notification fails
  due to authentication failure. This change is to capture some logs.

with @mrlumbu

### CODE REVIEW

File: src/main/java/uk/gov/pay/connector/auth/SmartpayAuthenticator.java [hunk 1]
```diff
@@ -4,19 +4,27 @@
...
     @Override
     public Optional<BasicAuthUser> authenticate(BasicCredentials basicCredentials) throws AuthenticationException {
-        return creds.asBasicCredentials().equals(basicCredentials)
-               ? Optional.of( new BasicAuthUser(creds.getUsername()) )
-               : Optional.absent();
+        if (creds.asBasicCredentials().equals(basicCredentials)) {
+            return Optional.of(new BasicAuthUser(creds.getUsername()));
+        }
+
+        logger.error(format("Authentication failure: failed for smartpay username %s", basicCredentials));
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/auth/SmartpayAuthenticator.java]: Do we want to print all the credentials here? The maximum allowed should be the username.

File: src/main/java/uk/gov/pay/connector/auth/SmartpayAccountSpecificAuthenticator.java [hunk 2]
```diff
@@ -25,9 +30,11 @@ public SmartpayAccountSpecificAuthenticator(GatewayAccountDao gatewayAccountDao,
 
         return gatewayAccountDao.findByNotificationCredentialsUsername(basicCredentials.getUsername())
                 .filter((gatewayAccountEntity) -> matchCredentials(basicCredentials, gatewayAccountEntity))
-                .map(gatewayAccountEntity -> gatewayAccountEntity.getNotificationCredentials().toBasicAuthUser())
-                .map(Optional::fromNullable)
-                .orElseGet(Optional::absent);
+                .map(gatewayAccountEntity -> Optional.fromNullable(gatewayAccountEntity.getNotificationCredentials().toBasicAuthUser()))
+                .orElseGet(() -> {
+                    logger.error(format("Authentication failure: failed for smartpay username %s", basicCredentials));
```
Comment (PauloPortugal) [inline src/main/java/uk/gov/pay/connector/auth/SmartpayAccountSpecificAuthenticator.java]: Do we want to print all the credentials here? The maximum allowed should be the username.

**Review by PauloPortugal — COMMENTED** [review]
I cannot see this in _PP-719_ the acceptance criteria to log the username/credentials.
If these are really required then we need to update the ticket in JIRA and we should not print all the credentials.
The question begs to be asked? Should we be logging usernames? If we get hacked and someone downloads all are logs or `grep` them then we will have a few accounts that could be hacked in a brute force manner.

File: src/main/java/uk/gov/pay/connector/auth/SmartpayAuthenticator.java (hunk 1 above)
Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/auth/SmartpayAuthenticator.java]: The toString() of the BasicCredentials class masks the password. It only shows the username.

File: src/main/java/uk/gov/pay/connector/auth/SmartpayAccountSpecificAuthenticator.java (hunk 2 above)
Comment (imranbohoran) [reply src/main/java/uk/gov/pay/connector/auth/SmartpayAccountSpecificAuthenticator.java]: The toString() of the BasicCredentials class masks the password. It only shows the username.

File: src/main/java/uk/gov/pay/connector/auth/SmartpayAuthenticator.java (hunk 1 above)
Comment (PauloPortugal) [reply src/main/java/uk/gov/pay/connector/auth/SmartpayAuthenticator.java]: :+1:

File: src/main/java/uk/gov/pay/connector/auth/SmartpayAccountSpecificAuthenticator.java (hunk 2 above)
Comment (PauloPortugal) [reply src/main/java/uk/gov/pay/connector/auth/SmartpayAccountSpecificAuthenticator.java]: :+1:

**Comment by PauloPortugal** [comment]
:+1: Spoke with Imran about logging the `username` and at the moment we do not have another option bu to do this. There is a plan to fix this in the near future.

**Pull request merged**